rust-runtime/aws-smithy-http-server/src/protocols.rs

```diff
@@ -7,8 +7,8 @@
 use crate::rejection::{ContentTypeRejection, MimeParsingFailed, MissingJsonContentType};
 use axum_core::extract::RequestParts;
 
-/// Validate that the request had the standard JSON content-type header.
-pub fn check_json_content_type<B>(req: &RequestParts<B>) -> Result<(), ContentTypeRejection> {
+/// Validate that the request had the standard JSON/XML content-type header.
+pub fn check_content_type<B>(req: &RequestParts<B>) -> Result<(), ContentTypeRejection> {
```

Comment:
I wasn't sure what to do here - I changed it to `check_content_type` and allow json or xml, but the error is `MissingJsonContentType`. Should I change it somehow?

Reply:
The original idea was to have multiple `check_<type>_content_type()` functions and to abstract the call to it inside the protocol implementation.

We are at a very early stage of development for the server codegen and we are working to finish the support of [restJson1](https://awslabs.github.io/smithy/1.0/spec/aws/aws-restjson1-protocol.html). Because of this part of the codebase is not as generic as I would like it to be.

What can be done is to have a switch matching the protocol type inside `ServerHttpProtocolGenerator` and depending on the protocol it will produce the right `check_<type>_content_type()` identifier:

```kotlin
fun getContentTypeCheck(): String {
      when (codegenContext.protocol) {
        RestJson1Trait.ID -> {
            return "check_json_content_type"
        }
        RestXmlTrait.ID -> {
            return "check_xml_content_type"
        }
        else -> {
            TODO("Protocol ${codegenContext.protocol} not supported yet")
        }
    }
}    
```

Later on

```kotlin
...
    val contentTypeCheck = getContentTypeCheck()
    rustTemplate(
        """
           let body = request.take_body().ok_or(#{SmithyHttpServer}::rejection::BodyAlreadyExtracted)?;
           let bytes = #{Hyper}::body::to_bytes(body).await?;
           if !bytes.is_empty() {
               #{SmithyHttpServer}::protocols::${contenTypeCheck}(request)?;
               input = #{parser}(bytes.as_ref(), input)?;
           }
        """,
        *codegenScope,
        "parser" to parser,
    )
...
```

I am suggesting this to unblock your PR. I have opened an issue #987 to further extract this logic inside the Protocol logic that can be addressed later.

Notes: check_json_content_type() is called [here](https://github.com/awslabs/smithy-rs/blob/main/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt#L522)

Reply:
I just added this to the PR - I hope this makes sense..